app/models/ability.rb

```diff
@@ -94,9 +94,11 @@ def initialize(member)
       cannot :destroy, OrderItem, :order => { :member_id => member.id, :completed_at => nil }
 
       # following/unfollowing permissions
-      can :create, Follow do |f|
-        !member.already_following?(f.followed) && f.followed_id != member.id
-      end
+      can :create, Follow
+      cannot :create, Follow, :followed_id => member.id # can't follow yourself
```

Comment:
You can't do it by typing in the address in your URL bar, as it requires 
a POST.  You have to use web dev tools to make the request.  If you do, 
for some reason, then the behaviour is undefined in the classic nasal 
demons (http://catb.org/jargon/html/N/nasal-demons.html) sense, but the 
worst that would happen would be for a duplicate of the HABTM 
relationship to be created and the name to be listed twice on your 
members/whoever/following page.

There is literally no way to prevent this when defining an ability, as 
in the line:

     can :create, Follow do |f|

... that "f" is never actually passed in, and the CanCan docs say that 
it will always return true.  I guess we could put some additional logic 
in the controller "create" method to return an error message if you play 
silly buggers with it.  But that then means our authorization checking 
is scattered across three different places in the code. Sigh.

A.

On 19/01/2015 11:41 pm, Cesy wrote:
>
> In app/models/ability.rb 
> <https://github.com/Growstuff/growstuff/pull/647#discussion-diff-23157734>:
>
> > @@ -94,9 +94,11 @@ def initialize(member)
> >        cannot :destroy, OrderItem, :order => { :member_id => member.id, :completed_at => nil }
> >
> >        # following/unfollowing permissions
> > -      can :create, Follow do |f|
> > -        !member.already_following?(f.followed) && f.followed_id != member.id
> > -      end
> > +      can :create, Follow
> > +      cannot :create, Follow, :followed_id => member.id # can't follow yourself
>
> What will now happen if you type in the address directly and try to 
> follow someone you're already following?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/Growstuff/growstuff/pull/647/files#r23157734>.
>

-- 
Alex "Skud" Bayley
skud@growstuff.org
http://growstuff.org/